### PR TITLE
Feature autocurrentpacket

### PR DESCRIPTION
**Pull request by hham**
Add `AutoCurrentPacket`. This is supposed to be the last piece of figure refactoring.

### CODE REVIEW

File: src/autowiring/AutoPacket.cpp [hunk 1]
```diff
@@ -226,8 +233,15 @@ void AutoPacket::UpdateSatisfactionUnsafe(std::unique_lock<std::mutex> lk, const
...
-  for (SatCounter* call : callQueue)
-    call->GetCall()(call->GetAutoFilter(), *this);
+  AutoPacket::SetCurrent(shared_from_this());
+  try {
+    for (SatCounter* call : callQueue)
+      call->GetCall()(call->GetAutoFilter(), *this);
+  } catch(...) {
+    AutoPacket::SetCurrent(NULL);
+    throw;
+  }
```
Comment (codemercenary) [inline src/autowiring/AutoPacket.cpp]: Can you use RAII here instead of a try-catch block?

File: src/autowiring/AutoPacket.cpp [hunk 2]
```diff
@@ -226,8 +233,15 @@ void AutoPacket::UpdateSatisfactionUnsafe(std::unique_lock<std::mutex> lk, const
...
+  AutoPacket::SetCurrent(NULL);
```
Comment (codemercenary) [inline src/autowiring/AutoPacket.cpp]: Use `nullptr`, never `NULL`.

File: src/autowiring/AutoPacket.cpp [hunk 3]
```diff
@@ -7,12 +7,19 @@
...
 #include <sstream>
 #include RVALUE_HEADER
 
 using namespace autowiring;
 
+/// <summary>
+/// A pointer to the current AutoPacket, specific to the current thread.
+/// </summary>
+/// <remarks>
+static thread_specific_ptr<std::weak_ptr<AutoPacket>> autoCurrentPacket;
```
Comment (codemercenary) [inline src/autowiring/AutoPacket.cpp]: Use a dumb pointer, here, not a weak pointer.

File: src/autowiring/AutoPacket.cpp (hunk 1 above)
Comment (hham) [inline src/autowiring/AutoPacket.cpp]: Yes, weak_ptr might be a over-kill. Done.

File: src/autowiring/AutoPacket.cpp (hunk 2 above)
Comment (hham) [inline src/autowiring/AutoPacket.cpp]: Done.

File: src/autowiring/AutoPacket.cpp (hunk 3 above)
Comment (hham) [inline src/autowiring/AutoPacket.cpp]: Yes, weak_ptr might be a over-kill. Done.

File: autowiring/AutoCurrentPacket.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+class AutoCurrentPacket
+{
+public:
+  AutoCurrentPacket(AutoPacket& apkt) {
+    AutoPacket::SetCurrent(apkt.shared_from_this());
+  };
+  ~AutoCurrentPacket(void) {
+    AutoPacket::SetCurrent(nullptr);
+  };
+};
```
Comment (codemercenary) [inline autowiring/AutoCurrentPacket.h]: Can you change the name of this class to `AutoCurrentPacketPusher` and put it in the `autowiring` namespace?

File: src/autowiring/AutoPacket.cpp [hunk 5]
```diff
@@ -7,12 +8,19 @@
...
 #include <algorithm>
 #include <sstream>
 #include RVALUE_HEADER
 
 using namespace autowiring;
 
+/// <summary>
+/// A pointer to the current AutoPacket, specific to the current thread.
+/// </summary>
+/// <remarks>
```
Comment (codemercenary) [inline src/autowiring/AutoPacket.cpp]: Missing close tag

File: autowiring/AutoCurrentPacket.h (hunk 4 above)
Comment (hham) [inline autowiring/AutoCurrentPacket.h]: Done.

File: src/autowiring/AutoPacket.cpp (hunk 5 above)
Comment (hham) [inline src/autowiring/AutoPacket.cpp]: Done.

**Pull request merged**